pkg/defkinds/nodejs/builder.go

```diff
@@ -0,0 +1,299 @@
+	}
+
+	return state, img, nil
+}
+
+func (h *NodeJSHandler) buildNodeJS(
+	ctx context.Context,
+	def Definition,
+	stageDef StageDefinition,
+	buildOpts builddef.BuildOpts,
+) (llb.State, *image.Image, error) {
+	state := llbutils.ImageSource(def.Locks.BaseImage, true)
+	baseImg, err := image.LoadMeta(ctx, def.Locks.BaseImage)
+	if err != nil {
+		return state, nil, xerrors.Errorf("loading %q metadata: %w", def.Locks.BaseImage, err)
+	}
+
+	img := image.CloneMeta(baseImg)
+	img.Config.Labels[builddef.ZbuildLabel] = "true"
+
+	state, err = llbutils.InstallSystemPackages(state, llbutils.APT, stageDef.Locks.SystemPackages)
+	if err != nil {
+		return state, img, xerrors.Errorf("failed to add \"install system pacakges\" steps: %w", err)
+	}
+
+	state = llbutils.CopyExternalFiles(state, stageDef.ExternalFiles)
+	state = llbutils.Mkdir(state, "1000:1000", "/app")
+	state = state.User("1000")
+	state = state.Dir("/app")
+
+	state = h.globalPackagesInstall(stageDef, state, buildOpts)
+
+	if *stageDef.Dev == false {
+		state = h.yarnInstall(stageDef, state, buildOpts)
+		state = h.copySourceDirs(stageDef, state, buildOpts)
+		state = h.build(stageDef, state)
+	}
+
+	setImageMetadata(stageDef, state, img)
+
+	return state, img, nil
+}
+
+func setImageMetadata(stageDef StageDefinition, state llb.State, img *image.Image) {
+	for _, dir := range stageDef.StatefulDirs {
+		fullpath := dir
+		if !path.IsAbs(fullpath) {
+			fullpath = path.Join("/app", dir)
+		}
+
+		img.Config.Volumes[fullpath] = struct{}{}
+	}
+
+	if *stageDef.Healthcheck {
+		img.Config.Healthcheck = &image.HealthConfig{
+			Test:     []string{"CMD", "http_proxy= test \"$(curl --fail http://127.0.0.1/_ping)\" = \"pong\""},
+			Interval: 10 * time.Second,
+			Timeout:  1 * time.Second,
+			Retries:  3,
+		}
+	}
+
+	nodeEnv := "development"
+	if *stageDef.Dev == false {
+		nodeEnv = "production"
+	}
+
+	img.Config.User = "1000"
+	img.Config.WorkingDir = "/app"
+	img.Config.Env = []string{
+		"PATH=" + getEnv(state, "PATH"),
+		"NODE_ENV=" + nodeEnv,
+	}
+	now := time.Now()
+	img.Created = &now
+
+	if stageDef.Command != nil {
+		img.Config.Cmd = *stageDef.Command
+	}
+}
+
+func (h *NodeJSHandler) buildWebserver(
+	ctx context.Context,
+	def Definition,
+	state llb.State,
+	buildOpts builddef.BuildOpts,
+) (llb.State, *image.Image, error) {
+	webserverHandler, err := registry.FindHandler("webserver")
+	if err != nil {
+		return state, nil, err
+	}
+	webserverHandler.WithSolver(h.solver)
+
+	webserverLocks, err := def.Locks.Webserver.RawLocks()
+	if err != nil {
+		return state, nil, err
+	}
+
+	newOpts := buildOpts
+	newOpts.Def = &builddef.BuildDef{
+		Kind:      "webserver",
+		RawConfig: def.Webserver.RawConfig(),
+		RawLocks:  webserverLocks,
+	}
+	newOpts.Source = &state
+
+	return webserverHandler.Build(ctx, newOpts)
+}
+
+func getEnv(src llb.State, name string) string {
+	val, _ := src.GetEnv(name)
+	return val
+}
+
+func (h *NodeJSHandler) globalPackagesInstall(
```

Comment:
why is there a need to install global packages ?

shouldn't such packages be part of the dev requirements ?

Reply:
I think it's quite common to install some dependencies like code generators as global dependencies (eg. https://github.com/api-platform/demo/blob/master/client/Dockerfile#L12). Moreover these deps are always installed in dev env, whereas usual npm deps are not.

Reply:
To say it another way, I think some people might want to use global packages in their images. As zbuild aims at making really easy for end users to create container images for their projects, I think the best is to not prevent people from doing what they're used to and to not impose such limitation :wink:

Reply:
> Moreover these deps are always installed in dev env, whereas usual npm deps are not.

I'm not sure to follow you there :thinking: 

NPM states that the deps required at runtime should go in the `deps` section, and all other deps should go in the `dev-deps` section.
The deps installed globally will be used by the project (for build reasons, or any other dev task), so it would make sense to specify them in the dev deps of the package.json file (i.e. tighten to the app) instead of in the dockerfile (i.e. tighten to the system packages) IMHO :thinking: 

Maybe you had an other use case in mind ?

It may not be a blocking discussion there, it's just that I don't see the benefit of such global package task for now ^^

Reply:
> To say it another way

I haven't see this post before replying ^^.
That's probably not the way I'll declare js deps that's OK :+1: